### PR TITLE
[MPS] Fix bfloat16 nextafter flushing subnormals to zero

### LINKED ISSUE #190740: [MPS] nextafter returns zero instead of the smallest bfloat16 subnormal when stepping away from zero
## Bug description

On MPS, `torch.nextafter` with bfloat16 returns zero when stepping away from zero: the result should be the smallest bf16 subnormal (`9.18e-41`), matching CPU. Steps from any nonzero value are correct, so this is specifically the zero -> subnormal case.

```python
import torch
for pair in [(0.0, 1.0), (0.0, -1.0), (-0.0, 1.0)]:
    x = torch.tensor([pair[0]], dtype=torch.bfloat16)
    y = torch.tensor([pair[1]], dtype=torch.bfloat16)
    print(pair, "cpu:", torch.nextafter(x, y).item(),
          "mps:", torch.nextafter(x.to("mps"), y.to("mps")).cpu().item())
```

Output on main (2.14.0a0, M5 Max, macOS 26):

```
(0.0, 1.0)  cpu: 9.183549615799121e-41  mps: 0.0
(0.0, -1.0) cpu: -9.183549615799121e-41 mps: -0.0
(-0.0, 1.0) cpu: 9.183549615799121e-41  mps: 0.0
```

Nonzero steps are exact on the same build (`nextafter(1, 2) == 1.0078125` etc.), and float32/float16 zero cases are correct. `test_nextafter` in test_mps.py fails on current main because of this (the `nextafter(0, 1) > 0` check).

## What this is not

- Not hardware flush-to-zero: a Metal shader that stores `as_type<bfloat>(ushort(1))` and even adds `bfloat(0.0)` to it preserves the `0x1` bit pattern on the same GPU.
- Not the pre-#190481 bug returning the input unchanged for all bf16 steps; nonzero steps work, so the open-coded bfloat functor from #190481 is being invoked.

## Suspected area

The bfloat `nextafter_functor` from #190481 handles `from == 0` explicitly (returns `as_type<bfloat>(sign | 1)`), yet exactly that case loses the subnormal, while the increment/decrement cases in the same functor work. Something between the functor's return value and the output store drops bf16 subnormals for this branch. #183055 reworked the binary tensor iterator (including `opmath_t` handling and store paths) after #190481 landed, so the interaction between the two is the natural first place to bisect; I have not pinned the exact mechanism.

## Versions

Observed on main (2.14.0a0, source build at 5862b066-era) on Apple M5 Max, macOS 26. CPU correct on the same build.

*This report was authored with assistance from Claude; the repro, controls, and FTZ shader probe were human-reviewed.*

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by SamDJ101**
Fixes #190740

### Summary

Noticed #190740 and looked into whether working on the raw bit patterns instead of bfloat values would fix it. It does.

`torch.nextafter` destroyed every bfloat16 subnormal on MPS, returning zero instead of the correct bit pattern. `test_nextafter` in `test_mps.py` fails on main on this hardware because of it. The bfloat overload now computes on sign-magnitude bits.

### Motivation

Two Metal behaviors combine here. Both are permitted by the Metal Shading Language specification, which allows denormals to be FTZ, so this needs a workaround in the kernel.

- bfloat16 comparisons treat subnormals as zero. `from == 0` is true and `from > 0` is false for a nonzero bf16 subnormal, so the old functor took its `from == 0` branch for every subnormal input, regardless of the argument.
- The `float -> bfloat` convert (a single `fptrunc`) flushes subnormal results, so the value that branch produced was lost on the way out. Measured over the full input space: with a nonzero bf16 subnormal as `from`, the old kernel returned ±0 in all cases.

### Changes

- Add `nextafter_bfloat_bits()`. This computes nextafter over raw sign-magnitude bits: zero detected by bit pattern, ordering via explicit sign-magnitude comparison, stepping via integer increment/decrement. NaN is detected by bit pattern but still returns `from + to`, matching the previous kernel and the CPU NaN payload exactly.
- Reduce the bfloat `operator()` to a single return that converts the resulting bits once.
- Leave the `float` and `half` overloads unchanged.
- Add bfloat16 subnormal sample inputs to the `nextafter` OpInfo, and compare bf16 exactly in `_compute_tolerances`. A flushed subnormal differs by ~1e-41, far inside the default bf16 tolerance of 0.016, so a tolerance-based comparison cannot see it.

### Test Plan

- Tested on Apple M4 Pro, macOS 27.0. Looking forward for CI results.
- Verified bit-exact against CPU over all 2^32 `(from, to)` bfloat16 pairs through the built runtime: 0 mismatches, NaN payloads included. On main the same sweep gives ~33M (0.78%) mismatches, and essentially every pair involving a subnormal is wrong (16,842,240 of 16,842,242).
- CPU was itself cross-checked against an independent IEEE implementation over 917k pairs with no mismatches, and agrees with the hand-written expected values in `test_nextafter_bfloat16`.
- No measurable performance change.

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/BinaryKernel.metal [hunk 1]
```diff
@@ -258,32 +258,45 @@ struct hermite_polynomial_he_functor {
   }
 };
 
+// Metal flushes bf16 subnormals both in float comparisons and in the
+// float->bfloat convert, so nextafter is computed on sign-magnitude bits and
+// materialized as a bfloat only at the functor's single return.
+inline ushort nextafter_bfloat_bits(const bfloat from, const bfloat to) {
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/kernels/BinaryKernel.metal]: STL next after doesn't work here?? ;-;

File: aten/src/ATen/native/mps/kernels/BinaryKernel.metal [hunk 2]
```diff
@@ -258,32 +258,45 @@ struct hermite_polynomial_he_functor {
...
+    return ushort(ut | 0x0040);
+  }
+  if (uf == ut) {
+    return ut;
+  }
+  if (af == 0 && at == 0) {
+    return ut;
+  }
+  if (af == 0) {
+    return ushort((ut & 0x8000) | 1);
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/kernels/BinaryKernel.metal]: Some of these magic byte values would benefit from a constexpr named variable or an explicit bit shift in construction to show what it's trying to do.

File: aten/src/ATen/native/mps/kernels/BinaryKernel.metal (hunk 1 above)
Comment (SamDJ101) [reply aten/src/ATen/native/mps/kernels/BinaryKernel.metal]: No bfloat overload unfortunately. `metal_math` only declares `nextafter` for `double`, `float` and `half`.
`::metal::nextafter(bf, bf)` widens its arguments and returns a `float`

File: aten/src/ATen/native/mps/kernels/BinaryKernel.metal (hunk 2 above)
Comment (SamDJ101) [reply aten/src/ATen/native/mps/kernels/BinaryKernel.metal]: Done in 0562af2

**Review by Skylion007 — APPROVED** [review]

File: test/test_mps.py [hunk 3]
```diff
@@ -15634,6 +15634,15 @@ def test_nextafter(self, device="mps"):
...
 
+    def test_nextafter_bfloat16_subnormals(self, device="mps"):
+        mags = torch.arange(0, 128, dtype=torch.int16).view(torch.bfloat16)
+        x_cpu = torch.cat([mags, -mags])
+        for to in [1.0, -1.0, 0.0]:
+            y_cpu = torch.full_like(x_cpu, to)
+            ref = torch.nextafter(x_cpu, y_cpu).view(torch.int16)
+            res = torch.nextafter(x_cpu.to(device), y_cpu.to(device)).cpu().view(torch.int16)
+            self.assertEqual(res, ref, atol=0, rtol=0)
+
```
Comment (Isalia20) [inline test/test_mps.py]: This test is not MPS specific, can you please move this test to OpInfo tests or wherever you dim appropriate so we can test this behaviour across different devices.

**Review by Isalia20 — DISMISSED** [review]

File: test/test_mps.py (hunk 3 above)
Comment (SamDJ101) [reply test/test_mps.py]: Thanks for the insight. I moved the test to the OpInfo in 94fa4cd8105

However, Sample inputs alone weren't enough. `test_output_match` compares numerically, and a flushed subnormal differs from the correct value by ~1e-41 which is far inside the default bf16 tolerance of 0.016. I've added a bf16-exact entry for `nextafter` in `_compute_tolerances` and the test in `test_binary_ufuncs.py` already compares it with `atol=0, rtol=0`.

Also found out that `test_nextafter_bfloat16` in `test_binary_ufuncs.py` already asserts these exact cases, but has never run on MPS, the file passes `allow_xpu=True` to `instantiate_device_type_tests` and not `allow_mps`. Enabling it surfaces an additional 11k MPS tests of which 1,086 fail on existing gaps (float64 and other logical ops)

File: test/test_mps.py [hunk 4]
```diff
@@ -16149,6 +16149,10 @@ def _assert_random_op_match(self, mps_out, cpu_out):
         self.assertEqual(mps_out.layout, cpu_out.layout)
 
     def _compute_tolerances(self, op, dtype):
+        # nextafter is defined on bit patterns, so bf16 must match exactly:
+        # a flushed subnormal differs by ~1e-41, far inside the default tolerance.
+        if op.name == "nextafter" and dtype == torch.bfloat16:
+            return (0, 0)
```
Comment (Isalia20) [inline test/test_mps.py]: This isn't really needed is it? Default tolerances should still pass

Comment (SamDJ101) [reply test/test_mps.py]: I beleive It's needed, this is because default tolerances pass whether the bug is present or not, so without this entry the subnormal samples can't fail.

Yes, its needed since `(0, 0)` is stricter than the default. nextafter returns the next representable value, so there is exactly one correct answer rather than an approximation to be tolerated. That is why `test_nextafter` in `test_binary_ufuncs.py` already compares it with `atol=0, rtol=0` on CPU.


Just verified here below:

```python
import torch
from torch.testing._comparison import default_tolerances

correct = torch.nextafter(torch.tensor([0.0], dtype=torch.bfloat16),
                          torch.tensor([1.0], dtype=torch.bfloat16)).item()
buggy = 0.0                                   # what the unpatched MPS kernel returns
rtol, atol = default_tolerances(torch.bfloat16)
diff = abs(correct - buggy)
allowance = atol + rtol * abs(correct)

print(f"diff= {diff}")
print(f"allowance = {allowance}")
print(f"considered close? = {diff <= allowance}")
```

```
diff= 9.183550e-41
allowance = 1.000000e-05
considered close? = True
```

File: test/test_mps.py (hunk 3 above)
Comment (Skylion007) [reply test/test_mps.py]: I think there a seperate OpInfo for MPS shim for this reason? It probably has a decorate failure somewhere

File: aten/src/ATen/native/mps/kernels/BinaryKernel.metal [hunk 5]
```diff
@@ -258,32 +258,49 @@ struct hermite_polynomial_he_functor {
   }
 };
 
+// bfloat16 layout: 1 sign bit, 8 exponent bits, 7 mantissa bits.
+constexpr constant ushort kBFloatSignMask = 1 << 15; // 0x8000
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/kernels/BinaryKernel.metal]: Don't do this. Use metal::numerics_limits to parameterize thus using the scalar_t?

Comment (SamDJ101) [reply aten/src/ATen/native/mps/kernels/BinaryKernel.metal]: Parameterized off numeric_limits in b7d502278b1. However, numeric_limits<bfloat>::denorm_min() can't be used here, Metal defines it as BFLT_MIN, the same as min(), so it returns 0x0080 (the smallest normal) rather than 0x0001

File: aten/src/ATen/native/mps/kernels/BinaryKernel.metal [hunk 6]
```diff
@@ -258,11 +258,15 @@ struct hermite_polynomial_he_functor {
...
-// bfloat16 layout: 1 sign bit, 8 exponent bits, 7 mantissa bits.
-constexpr constant ushort kBFloatSignMask = 1 << 15; // 0x8000
-constexpr constant ushort kBFloatMagMask = (1 << 15) - 1; // 0x7FFF
-constexpr constant ushort kBFloatInf = 0xFF << 7; // 0x7F80, exponent all ones
-constexpr constant ushort kBFloatQuietBit = 1 << 6; // 0x0040, top mantissa bit
+// Metal defines denorm_min() as BFLT_MIN, so it
+// reports the smallest normal (0x0080) instead of 0x0001.
+using bfloat_bits_t = ushort;
+constexpr constant bfloat_bits_t kBFloatSignMask = bfloat_bits_t(1)
+    << (sizeof(bfloat_bits_t) * 8 - 1);
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/kernels/BinaryKernel.metal]: The size of the short type has nothing do with bfloat16?

Comment (SamDJ101) [reply aten/src/ATen/native/mps/kernels/BinaryKernel.metal]: Fixed in 4f515462a1e. The shift now derives from `sizeof(bfloat) * CHAR_BIT` rather than the storage type, matching how `_fp_encoding_traits_impl` computes `bit_width` in `metal_limits`. I used tge previous form because `ushort` and `bfloat` are both 2 bytes.

Did some refactor to scope the constants inside.

I evaluated `_fp_encoding_traits<bfloat>`, which exposes `sign_mask = 0x8000` and `inf_mask = 0x7F80` directly. I did not use it as it might have no stability guarantee across toolchain versions (not sure here).

Happy to switch if you prefer it that way

Comment (SamDJ101) [reply aten/src/ATen/native/mps/kernels/BinaryKernel.metal]: Updated the test cases as well.

**Review by malfet — APPROVED** [review]
LGTM, a bit too verbose, but correct

File: test/test_mps.py [hunk 7]
```diff
@@ -16149,6 +16149,10 @@ def _assert_random_op_match(self, mps_out, cpu_out):
         self.assertEqual(mps_out.layout, cpu_out.layout)
 
     def _compute_tolerances(self, op, dtype):
+        # nextafter is defined on bit patterns, so bf16 must match exactly:
+        # a flushed subnormal differs by ~1e-41, far inside the default tolerance.
+        if op.name == "nextafter" and dtype == torch.bfloat16:
```
Comment (malfet) [inline test/test_mps.py]: But it should be like that for all dtypes, not just blofatt16...
